### PR TITLE
Fix: Issues with upgrading colony version

### LINKED ISSUE #1677: Issues with upgrading colony version
## Summary

Prior to deploying the beta, we had hope to do a contract release. We didn't manage to do so, so it's important that colony upgrades work as expected for our existing beta users when the next (and future) release happens.

## How to test

Currently, this is not easy - this is something I want to change. But for now, this is what it is. If you try to reproduce locally and find it not working as described, please reach out to me and we'll get to the bottom of it (and I will update these instructions).

1. Check out a version of the CDapp repository with an alternative `create-data` script. Once you've cloned the repository, `git checkout fix/deploy-old-version`
2. In a new directory, checkout the previous version of colonyNetwork (`git clone git@github.com:JoinColony/colonyNetwork` and then `git checkout glwss4`
3. Make sure you've run `npm i` in both directories
4. Run the dev environment as usual in the CDapp (`npm run dev`), and wait until it has finished setting everything up. Run `npm run webpack` in another window as usual.
5. Deploy the old colony version resolver with the following command in the colonyNetwork repository: 
```
npx truffle exec --network development ./scripts/deployOldUpgradeableVersionTruffle.js --interfaceName IColony --implementationNames ColonyArbitraryTransaction,ColonyDomains,ColonyExpenditure,ColonyFunding,ColonyPayment,ColonyRewards,ColonyRoles,ColonyTask,ContractRecovery,Colony
```
It will output the address of the new resolver. I would expect it to be `0x0F1161B679D63434B6154cFa7015EC921a2f7924` though as this issue ages and the exact transactions that happen change, that will change.
6. Add the old version resolver to the deployed colony network. The colony network address can be found at http://localhost:3006/etherrouter-address.json, and is probably `0x5CC4a96B08e8C88f2c6FC5772496FeD9666e4D1F`. The following instructions assume these addresses - if they've changed, then you will need to update the instructions appropriately
7. In a truffle console, accessed from the network repository with `npx truffle console`:
```
cn = await IColonyNetwork.at("0x5CC4a96B08e8C88f2c6FC5772496FeD9666e4D1F")
cn.getMetaColony()
_Expected output: `0x1133560dB4AebBebC712d4273C8e3137f58c3A65`_
mc = await IMetaColony.at("0x1133560dB4AebBebC712d4273C8e3137f58c3A65")
mc.addNetworkColonyVersion(13, "0x0F1161B679D63434B6154cFa7015EC921a2f7924")
````
8. You have now added a v13 resolver to the local development environment. Now run, from the CDapp repo `node ./scripts/create-data.js --coloniesCount 4`
This will deploy four colonies. The fourth colony will be deployed at version 13.

# Issues 

## In up-to-date colonies
screenshot-2024-01-09_15-33-13
Planet Express is v14, but there is still an 'upgrade colony' link in this menu. If you click it, it pops out...
screenshot-2024-01-09_15-33-18
Aside from the 'v's being in different places here in the description, `v15` doesn't exist! Clicking 'upgrade' does nothing, which might indicate the CDapp recognises this to some extent, or it might indicate that the whole form is broken...

## In old colonies
 
'Oscorp Industries' is an old colony, if you've used my adjusted script, so navigate there.

screenshot-2024-01-09_15-38-07

I think the prose here is confusing - I would expect something like "Your colony's current version" and "Latest network version". But we have the banner at the top, and we have the 'Upgrade Now' button here which is disabled in an up-to-date colony, so that is working correctly. 

Clicking 'Upgrade Now' does nothing. Click 'Upgrade' in the banner gives the pop-up from the side to execute an action, but clicking 'Upgrade' there then does nothing (no network requests, no errors in the console).

## Extras
I am currently only testing these with permissions, but upgrading should also be possible via motions

### PR DESCRIPTION
**Pull request by iamsamgibbs**
## How to Test

Requires changes to your dev environment to have a Colony running on version 13. Please follow the steps in #1677 (You will also need to copy over this script into ColonyNetwork as it is missing from the `glwss4` branch: https://github.com/JoinColony/colonyNetwork/blob/develop/scripts/deployOldUpgradeableVersionTruffle.js)

## Description

If already on the latest colony version, the action sidebar will now show an info banner and disable the submit button when the 'upgrade colony version' action type is selected. Metadata has also been adjusted to: `{Upgrade Colony version to v{newVersion} by {initiator}}` and "latest version" in the table, as discussed with Arren.

Screenshot 2024-01-18 at 13 53 41

On the advanced settings page, the current colony version is now on the left and latest contract version on the right. Text has been updated from "Last version" to "Latest version".

Screenshot 2024-01-18 at 13 48 25

"Upgrade now" button now opens the action sidebar with the upgrade colony action selected.

Screenshot 2024-01-18 at 13 48 56

"Upgrade Colony version" link in the admin menu should remain as discussed with Arren.

**Changes** 🏗

* If already on the latest colony version, action sidebar now shows an info banner
* If already on the latest colony version, action sidebar submit button is disabled
* Action sidebar metadata adjusted to `{Upgrade Colony version to v{newVersion} by {initiator}}`
* In the action sidebar form: "New version" updated to "Latest version"
* On the advanced settings page, current colony version and latest contract version now in the correct positions, and "Last version" adjusted to "Latest version".
* "Upgrade now" button on advanced settings page now opens action sidebar

Resolves #1677

### CODE REVIEW

**Comment by rdig** [comment]
@area since you're the one that brought this up initially, I'd want you to also test it locally and check if it meets your requirements

File: src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/SidebarBanner.tsx [hunk 1]
```diff
@@ -88,6 +100,13 @@ export const SidebarBanner: FC = () => {
           </NotificationBanner>
         </div>
       ))}
+      {showVersionUpToDateNotification && (
+        <div className="mt-6">
+          <NotificationBanner icon="check-circle" status="success">
```
Comment (rdig) [inline src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/SidebarBanner.tsx:105]: Why is the status of `NotificationBanner` a plain string and not an `enum`.

_(This is not directed at you @iamsamgibbs, it's just me ranting about the implemetation)_

File: src/components/v5/common/ActionSidebar/partials/hooks.ts [hunk 2]
```diff
@@ -57,5 +61,9 @@ export const useSubmitButtonDisabled = () => {
     return true;
   }
 
+  if (selectedAction === ACTION.UPGRADE_COLONY_VERSION && !canUpgrade) {
+    return true;
+  }
+
```
Comment (rdig) [inline src/components/v5/common/ActionSidebar/partials/hooks.ts:67]: This is starting to drag with all the if cases (2 now, more on the way most likely), so please refactor this as a `switch` statement

**Review by rdig — APPROVED** [review]
Very nice work @iamsamgibbs :100: 

Tested this with actions, working perfectly, however motions were harder to test since I've ran into some issues _(which currently affect `master`)_

I've left a note about refactoring that submit button hook, but otherwise this is good to go.

Screenshot from 2024-01-19 18-31-44
Screenshot from 2024-01-19 18-31-57

File: src/components/frame/Extensions/pages/AdvancedPage/AdvancedPage.tsx [hunk 3]
```diff
@@ -48,8 +48,8 @@ const AdvancedPage: FC = () => {
         )}
       </div>
       <ColonyVersionWidget
-        currentVersion={colonyContractVersion}
-        lastVersion={version || 0}
+        currentVersion={version || 0}
```
Comment (jakubcolony) [inline src/components/frame/Extensions/pages/AdvancedPage/AdvancedPage.tsx:51]: Might not need the fallback anymore

**Review by jakubcolony — APPROVED** [review]
Great work @iamsamgibbs!

File: src/components/frame/Extensions/pages/AdvancedPage/AdvancedPage.tsx (hunk 3 above)
Comment (iamsamgibbs) [reply src/components/frame/Extensions/pages/AdvancedPage/AdvancedPage.tsx]: Removed this unnecessary fallback.

File: src/components/v5/common/ActionSidebar/partials/hooks.ts (hunk 2 above)
Comment (iamsamgibbs) [reply src/components/v5/common/ActionSidebar/partials/hooks.ts]: Refactored this as a switch statement.

File: src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/SidebarBanner.tsx (hunk 1 above)
Comment (iamsamgibbs) [reply src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/SidebarBanner.tsx:105]: Opened as new issue: #1770

**Review by rdig — DISMISSED** [review]

**Review by jakubcolony — DISMISSED** [review]

**Comment by rdig** [comment]
@iamsamgibbs can you please rebase this on the latest `master`. It brings in the motion fixes which'll make this easier to test for @area

**Comment by area** [comment]
Creating a motion (after adding the relevant extension) for a colony version upgrade does not appear to work cleanly. The motion is created, and exists, but when trying to see the details of the motion I get the below.

screenshot-2024-01-24_11-31-03

Having spoken to Raul, he thinks it's likely that there is a switch somewhere in the rendering specifically for motions, and there simply isn't an implementation for a motion that is to upgrade a colony (whereas there is for an action that upgrades a colony, and there are implementations for other motions).

@arrenv, is there a design for this somewhere (i.e. specifically a motion to upgrade a colony)?

**Review by rdig — CHANGES_REQUESTED** [review]
Checked this with Alex, and indeed, I can reproduce the upgrade via motions issues.

Reverted my approval until that is going to get implemented.

PS: I do realize that this increases the scope of this PR

**Comment by iamsamgibbs** [comment]
Added VersionUpgradeMotion as a supported action for the CompletedAction component so the details now show correctly when creating a version upgrade motion.

https://github.com/JoinColony/colonyCDapp/assets/34915414/d2527d7d-a47b-42a1-b72d-4a7ef6b2d7cf

(Unrelated(?) issue: description takes a couple of seconds to pop in - not sure if this happens on other motions?)

---

Also added 'Current version' and 'Latest version' to CompletedAction component which shows when created with permissions or as a motion.

https://github.com/JoinColony/colonyCDapp/assets/34915414/cf1e8d34-405b-4b37-a07a-f20cb5268544

(Unrelated(?) issue: when creating with permissions, title initally appears with the fallback before updating to the correct value. Description takes a couple of seconds to pop in.)

**Comment by area** [comment]
Functionality all appears there, both for motions and for doing the upgrade via permissions. For the component, I would not have 'Latest Version', which is only accurate if the colony in question is one version behind, and would be arguably misleading in the future. Perhaps 'Upgrade to Version'?

I did encounter some presumably undesirable behaviour, but probably worthy of its own issue and PR, rather than holding up this one, as it affects all motions:

* Create two motions to upgrade the colony, and move them both to the 'execution' step.
* The first one can be finalized
* The second one cannot (which is the correct contract behaviour), but clicking the 'Finalize' button appears to do nothing in the UI, and logs the `voting-execution-failed-not-one-week` error that occurs when (presumably) estimating gas to the console. Repeatedly clicking the button is possible, and has the same effect each time. I don't _believe_ this is actually broadcasting a transaction though, so users will not be losing gas fees, just getting confused.

**Review by area — DISMISSED** [review]

**Comment by iamsamgibbs** [comment]
Thanks for testing this! I have updated "Latest version" to "New version" as per the figma design.

Screenshot 2024-01-26 at 12 29 56

I was able to recreate the finalising issue and agree it should be raised as a new issue.

**Review by area — DISMISSED** [review]
:+1:

File: src/components/v5/shared/ColonyVersionWidget/ColonyVersionWidget.tsx [hunk 4]
```diff
@@ -14,12 +17,24 @@ const displayName = 'v5.ColonyVersionWidget';
...
   currentVersion,
 }) => {
   const { formatMessage } = useIntl();
   const isMobile = useMobile();
 
+  const {
+    actionSidebarToggle: [, { toggleOn: toggleActionSidebarOn }],
+  } = useActionSidebarContext();
+
+  const handleUpgradeColony = useCallback(
```
Comment (rdig) [inline src/components/v5/shared/ColonyVersionWidget/ColonyVersionWidget.tsx:30]: I don't think `useCallback` is really necessary here, as the toggle action does not benefit from being memoized

**Review by rdig — APPROVED** [review]
Looks good.

Tested both actions and motions and now works as expected. Nice work yet again Sam!

Screenshot from 2024-01-26 19-23-45
Screenshot from 2024-01-26 19-25-32
Screenshot from 2024-01-26 19-25-47

File: src/components/v5/shared/ColonyVersionWidget/ColonyVersionWidget.tsx (hunk 4 above)
Comment (iamsamgibbs) [reply src/components/v5/shared/ColonyVersionWidget/ColonyVersionWidget.tsx]: I've removed this now 👍

**Review by rdig — APPROVED** [review]
Re-approved. Nice work.

**Review by area — APPROVED** [review]

**Pull request merged**